power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -145,7 +145,31 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
 
     using OwnedUpdateDataset = std::tuple<std::vector<typename ComponentType::UpdateType>...>;
 
+    struct UpdateCompProperties {
+        std::string name{};
+        bool has_id{false};           // if the component has id
+        bool ids_all_na{false};       // if all ids are all NA
+        bool ids_part_na{false};      // if some ids are NA but some are not
+        bool uniform{false};          // if the component is uniform
+        bool is_columnar{false};      // if the component is columnar
+        bool ids_match{false};        // if the ids match
+        Idx elements_ps_in_update{0}; // count of elements for this component per scenario in update
+        Idx elements_in_base{0};      // count of elements for this component per scenario in input
+        bool no_id_col() const { return is_columnar && (!has_id || ids_all_na); }
+        bool no_id_row() const { return !is_columnar && (!has_id && ids_all_na); }
+        bool qualify_for_optional_id() const { return ids_match && ids_all_na && !ids_part_na; }
+        bool provided_ids_valid() const { return has_id && ids_match && !ids_all_na && !ids_part_na; }
+        bool is_independent() const { return qualify_for_optional_id() || provided_ids_valid(); }
+        Idx get_n_elements() const {
```

Comment:
```suggestion
        constexpr bool no_id_col() const { return is_columnar && (!has_id || ids_all_na); }
        constexpr bool no_id_row() const { return !is_columnar && (!has_id && ids_all_na); }
        constexpr bool qualify_for_optional_id() const { return ids_match && ids_all_na && !ids_part_na; }
        constexpr bool provided_ids_valid() const { return has_id && ids_match && !ids_all_na && !ids_part_na; }
        constexpr bool is_independent() const { return qualify_for_optional_id() || provided_ids_valid(); }
        constexpr Idx get_n_elements() const {
```

Reply:
These are runtime properties.

Reply:
After refactoring, yes.

Reply:
the properties are runtime but the functions are not. any function that does not strictly require access to the data shall be `constexpr`.

a good rule of thumb is: "if I were to provide hardcoded input values, should it be able to give me a `constexpr` output?"

`constexpr` means "if you can understand what the expression means at compile-time, and you deem it faster, please do".

what you are thinking of is forcing compile-time evaluation (`consteval`) or compile-time resolution with runtime initialization (`constinit`).

Reply:
Addressed.